railgun/internal/proxy/clientgo_cached_proxy_resolver.go

```diff
@@ -295,10 +302,10 @@ func (p *clientgoCachedProxyResolver) cacheDelete(cobj *cachedObject) error {
 	return cobj.err
 }
 
-// kongRootWithTimeout provides the root configuration from Kong, but uses a default timeout to avoid long waits if the Admin API
+// kongRootWithTimeout provides the root configuration from Kong, but uses a configurable timeout to avoid long waits if the Admin API
 // is not yet ready to respond. If a timeout error occurs, the caller is responsible for providing a retry mechanism.
 func (p *clientgoCachedProxyResolver) kongRootWithTimeout() (map[string]interface{}, error) {
-	ctx, cancel := context.WithTimeout(p.ctx, 3*time.Second)
+	ctx, cancel := context.WithTimeout(p.ctx, p.timeout)
 	defer cancel()
 	return p.kongConfig.Client.Root(ctx)
 }
```

Comment:
This patch looks good, thank you! :bow: 

Before we merge I would like to see a small unit test added (see `clientgo_cached_proxy_resolver_test.go`) which validates this. It could just be a couple of lines which set a microsecond timeout, uses the existing admin api mocking to ensure the response will not finish within that time, and validate the error. It's a small thing, but for a minuscule cost it could help protect against regressions in the future if a future contributor manages to accidentally break the timeout mechanism and doesn't realize it.

Reply:
@shaneutt, are there any references for this `timeout` test in the codebase?

Reply:
just a thought can I update the existing `TestCaching` test with the timeout assertions?

Reply:
@shaneut, I added the test but I am not sure why it is not passing. Need ur help on this 🙏

Reply:
sure, I'll take a look.

Reply:
I think https://github.com/Kong/kubernetes-ingress-controller/pull/1401#discussion_r647713112 might be related, try just sleeping in the response function and I expect this should work? :thinking:

Reply:
Yes, It should work but I debugged the test and found that the `endpoint` we are changing as per our requirements to the kong fake admin server is not taking up our change. so, it always goes to the default response.

Reply:
@shaneutt, is anything I missed?

Reply:
I'll pull this branch locally and take a look today :heavy_check_mark:

Reply:
Taking a look...

Reply:
I'm sorry, the problem is partially the lack of documentation (yet) for some of these mocking tools from the KTF. You have to mock the responses from the API using the `MockNextResponse()` method. The following diff can be applied to fix the tests and get a :heavy_check_mark: result:

```diff
diff --git a/railgun/internal/proxy/clientgo_cached_proxy_resolver_test.go b/railgun/internal/proxy/clientgo_cached_proxy_resolver_test.go
index c8c40ea1..687feef9 100644
--- a/railgun/internal/proxy/clientgo_cached_proxy_resolver_test.go
+++ b/railgun/internal/proxy/clientgo_cached_proxy_resolver_test.go
@@ -3,13 +3,13 @@ package proxy
 import (
 	"context"
 	"net/http"
-	"net/http/httptest"
 	"testing"
 	"time"
 
 	"github.com/google/uuid"
 	"github.com/kong/kubernetes-ingress-controller/pkg/store"
 	"github.com/kong/kubernetes-testing-framework/pkg/generators/k8s"
+	"github.com/kong/kubernetes-testing-framework/pkg/kong"
 	"github.com/stretchr/testify/assert"
 	corev1 "k8s.io/api/core/v1"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
@@ -142,11 +142,18 @@ func TestProxyTimeout(t *testing.T) {
 	defer cancel()
 	t.Log("configuring and starting a new proxy server")
 
+	// mock the next Admin API response (which will be / to get the root config) to ensure
+	// that it takes longer than the timeout we will set, in order to trigger the timeout.
+	fakeKongAdminAPI.MockNextResponse(kong.AdminAPIResponse{
+		Status:   http.StatusGatewayTimeout,
+		Body:     []byte{},
+		Callback: func() { time.Sleep(time.Millisecond * 30) },
+	})
+
+	// the timeout is shorter than the wait time for the http response, we should expect
+	// to see the the context deadline for the http response triggered.
 	timeout := time.Millisecond * 10
-	fakeKongAdminAPI.Endpoint = httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
-		time.Sleep(time.Millisecond * 30) // so we are sure we exceed the timeout
-	}))
-	fakeKongConfig.Client = fakeKongAdminAPI.KongClient
 	_, err := NewCacheBasedProxy(ctx, logger, fakeK8sClient, fakeKongConfig, "kongtests", false, mockKongAdmin, timeout)
 	assert.Error(t, err)
+	assert.Contains(t, err.Error(), "context deadline exceeded")
 }
```

Let me know if this is helpful and feel free to apply that diff verbatim.

Reply:
aah yes ! this is what i was checking as of now ( going through KTF codebase ). Thanks @shaneutt

Reply:
No problem, sorry for the delay :bow:

Reply:
no problem 😅 thanks for helping